waku/waku_sync/protocol.nim

```diff
@@ -0,0 +1,311 @@
+
+    return err("Sync request error: " & error)
+
+  return ok((hashes, peer))
+
+proc handleLoop(
+    self: WakuSync, conn: Connection
+): Future[Result[seq[WakuMessageHash], string]] {.async.} =
+  let (start, `end`) = calculateRange(self.relayJitter)
+
+  let initialized =
+    ?serverInitialize(self.storage, conn, self.maxFrameSize, start, `end`)
+
+  var sent = initialized
+
+  while true:
+    let received = ?await sent.listenBack()
+
+    trace "sync payload received",
+      server = self.peerManager.switch.peerInfo.peerId,
+      client = conn.peerId,
+      payload = received.payload
+
+    let reconciled = (?received.serverReconcile()).valueOr:
+      let completed = error # Result[Reconciled, Completed]
+
+      let hashes = await completed.serverTerminate()
+
+      return ok(hashes)
+
+    sent = ?await reconciled.send()
+
+    trace "sync payload sent",
+      server = self.peerManager.switch.peerInfo.peerId,
+      client = conn.peerId,
+      payload = reconciled.payload
+
+proc initProtocolHandler(self: WakuSync) =
+  proc handle(conn: Connection, proto: string) {.async, gcsafe, closure.} =
+    debug "sync session requested",
+      server = self.peerManager.switch.peerInfo.peerId, client = conn.peerId
+
+    let hashes = (await self.handleLoop(conn)).valueOr:
+      debug "sync session ended",
+        server = self.peerManager.switch.peerInfo.peerId, client = conn.peerId, error
+
+      #TODO send error code and desc to client
+      return
+
+    if hashes.len > 0 and self.transferCallBack.isSome():
+      let callback = self.transferCallBack.get()
+
+      (await callback(hashes, conn.peerId)).isOkOr:
+        debug "transfer callback failed", error
+
+    debug "sync session ended gracefully",
+      server = self.peerManager.switch.peerInfo.peerId, client = conn.peerId
+
+  self.handler = handle
+  self.codec = WakuSyncCodec
+
+proc new*(
+    T: type WakuSync,
+    peerManager: PeerManager,
+    maxFrameSize: int = DefaultMaxFrameSize,
+    syncInterval: timer.Duration = DefaultSyncInterval,
+    relayJitter: Duration = DefaultGossipSubJitter,
+    pruneCB: Option[PruneCallBack] = none(PruneCallback),
+    transferCB: Option[TransferCallback] = none(TransferCallback),
+): Result[T, string] =
+  let storage = Storage.new().valueOr:
+    return err("negentropy storage creation failed")
+
+  let sync = WakuSync(
+    storage: storage,
+    peerManager: peerManager,
+    maxFrameSize: maxFrameSize,
+    syncInterval: syncInterval,
+    relayJitter: relayJitter,
+    pruneCallBack: pruneCB,
+    pruneOffset: syncInterval div 2,
+    transferCallBack: transferCB,
+  )
+
+  sync.initProtocolHandler()
+
+  info "WakuSync protocol initialized"
+
+  return ok(sync)
+
+proc periodicSync(self: WakuSync, callback: TransferCallback) {.async.} =
+  debug "periodic sync initialized", interval = $self.syncInterval
+
+  while true:
+    await sleepAsync(self.syncInterval)
```

Comment:
Should it better be at the end of the loop?

Reply:
No? It would trigger a sync when initialized. We planned a feature for this purpose, some kind of Store resume.